earth_enterprise/src/common/khxml/khxml.cpp

```diff
@@ -96,271 +91,263 @@ class PercentError : public XmlParamsException
+                       "maxDOMHeapAllocSize=%zu\n"
+                       "maxDOMSubAllocationSize=%zu\n"
+                       "doPurge=%s\n"
+                       "purgeLevel=%d\n"
+                       "purgeThreshold=%zu\n"
+                       "deallocateAll=%s",
+             initialDOMHeapAllocSize,
+             maxDOMHeapAllocSize,
+             maxDOMSubAllocationSize,
+             (doPurge ? "true" : "false"),
+             purgeLevel,
+             purgeThreshold,
+             (deallocateAll ? "true" : "false"));
     }
     catch (const XMLException& toCatch)
     {
-        notify(NFY_FATAL, "Unable to initialize Xerces: %s",
-               FromXMLStr(toCatch.getMessage()).c_str());
+      notify(NFY_FATAL, "Unable to initialize Xerces: %s",
+             FromXMLStr(toCatch.getMessage()).c_str());
     }
+  }
 }
 
-static khMutexBase cacheLock = KH_MUTEX_BASE_INITIALIZER;
-static khMutexBase reinitLock = KH_MUTEX_BASE_INITIALIZER;
-
-// A read-write guard to aid in purging
-//
-// Before the creation of DOMDocuments and DOMLSParsers, a call to
-// reInitIfNeeded() is made where it is determined whether or not
-// to purge the XML cache.  There are 5 levels at which to purge
-// the cache: 1 being the most frequent, 5 being the least frequent
-// 
-// Every time a DOMDocument enters into a write operation
-// a "write lock" is made, which essentially is just keeping a count 
-// of write operations remaining. When it is determined it is time 
-// for a purge to occur, a "purge lock" is made.
-//
-// When a "purge lock" occurs, it waits for all "write locks" to finish
-// and then takes control. It will execute all purge operations and then
-// cede control  back to the control of write operations.
-//
-// TODO: generalize and move read-write lock to khThread
-
-static pthread_rwlock_t purge_lock = PTHREAD_RWLOCK_INITIALIZER;
-
-class XmlWriteLock
-{
-private:
-    pthread_rwlock_t& lock;
-public:
-    XmlWriteLock(pthread_rwlock_t& _lock) : lock(_lock)
-    {
-        pthread_rwlock_rdlock(&lock);
-    }
-    
-    ~XmlWriteLock()
-    {
-        pthread_rwlock_unlock(&lock);
-    }
-};
-
-class PurgeLock
-{
-private:
-    pthread_rwlock_t& lock;
-public:
-    PurgeLock(pthread_rwlock_t& _lock) : lock(_lock)
-    {
-        pthread_rwlock_wrlock(&lock);
-    }
-
-    ~PurgeLock()
-    {
-        pthread_rwlock_unlock(&lock);
-    }
-};
-
-void ReInitializeXerces()
-{
-    khLockGuard guard(reinitLock);
-    try
-    {
-        XMLPlatformUtils::Terminate();
-    }
-    catch(...)
-    {
-        notify(NFY_WARN, "Failure to terminate in ReInitializeXerces()");
+GEXMLObject::~GEXMLObject() {
+  khLockGuard guard(mutex);
+  --activeObjects;
+  // Terminate Xerces and clear all memory when the user says we can, the last
+  // object is destroyed, and the cache size is over the threshold.
+  if (doPurge && activeObjects == 0 && memoryManager.size() >= purgeThreshold) {
+    try {
+      XMLPlatformUtils::Terminate();
```

Comment:
Would it be potentially useful to have a debug message in here somewhere so we can see how often this is hitting?